scikit-learn/scikit-learn
# NOCATS: Categorical splits for tree-based learners (ctnd.)

This PR continues the work of #4899. For now I've merged the master into the PR, made it compile and make the tests run. There are several issues which need to be fixed. The list will be updated as I encounter them. Also, not all of these items are necessarily open, I have only collected them from the comments on the original PR, and need to make sure they're either already addressed or address them.

- merge master into the PR (__done__)
- sparse tests pass (__done__)
    - The code is supposed to be the same as the status quo implementation if categories are not passed. But right now the tests related to sparse data fail.
    - __EDIT__: The tests pass if we compare floats with `almost_equal`
- LabelEncoder -> CategoricalEncoder (__done__)
    - Preprocessing is not a part of NOCATS anymore.
- Is maximum random generations 20 or 40 (__done__)
    - It's actually 60
- Don't quantize features automatically (__done__)
    - Doesn't happen anymore: https://github.com/scikit-learn/scikit-learn/pull/4899#issuecomment-271504258
- check the category count limits for given data. (__done__)
- add a benchmark
   - __done__. Results: https://github.com/scikit-learn/scikit-learn/pull/12866#issuecomment-453856876
- add tests (right now only invalid input are tested)
   - `tree/tests` __done__
   - `ensemble/tests` __done__
- benchmark against master
- add an example with plots
- check numpy upgrade related issues (we've upgraded our numpy requirement in the meantime)
- run some benchmarks with a simple integer coding of the features (with arbitrary ordering)
- add cat_split to NODE_DTYPE once joblib.hash can handle it (padded struct)
    - joblib issue: joblib/joblib#826

Closes #4899

__Future Work__: These are the possible future work we already know of (i.e. outside the scope of this PR):

- Heuristic methods to allow fast Breiman-like training for multi-class classification
- export to graphviz
- One-hot emulation using the NOCATS machinery
- support sparse input
- handle categories as their unique valies instead of `[0, max(feature)]`
  - This is to be consistent with our encoders' behavior
  - moved this to future work per https://github.com/scikit-learn/scikit-learn/pull/12866#issuecomment-455021204


P.S. I moved away from "task list" due to the extremely buggy interface when used in combination with editing the post, which I'm extensively doing to keep it easy for us to keep up with the status.
